modelmapper/fetcher.py

```diff
@@ -0,0 +1,262 @@
+            raise ImportError('Please install Boto3')
+
+try:
+    import mmh3
+except ImportError:
+    class mmh3:
+        def hash(self, *args, **kwargs):
+            raise ImportError('Please install mmh3')
+
+import gzip
+import datetime
+import logging
+
+from collections import Mapping
+from botocore.client import Config
+
+from modelmapper.cleaner import Cleaner
+from modelmapper.slack import slack
+from modelmapper.misc import generator_chunker, generator_updater
+
+try:
+    from sqlalchemy.dialects.postgresql import insert
+except ImportError:
+    def insert(table):
+        raise ImportError('Please install SQLAlchemy')
+
+
+class Fetcher(Cleaner):
+    """
+    Subclass this for your data processing and define the BUCKET_NAME, RAW_KEY_MODEL and RECORDS_MODEL.
+
+    """
+
+    BUCKET_NAME = None
+    RAW_KEY_MODEL = None
+    RECORDS_MODEL = None
+    S3KEY_DATETIME_FORMAT = '%Y/%m/%Y_%m_%d__%H_%M_%S.gzip'
+    SQL_CHUNK_ROWS = 300
+    logger = logging.getLogger(__name__)
+
+    def __init__(self, *args, **kwargs):
+        self.JOB_NAME = self.__class__.__name__
+        self.DUMP_FILEPATH = f'/tmp/{self.JOB_NAME}_dump'
+        super().__init__(*args, **kwargs)
+
+    def get_client_data(self):
+        """
+        The client should return data that the importer will clean and import.
+        """
+        raise NotImplementedError('Please implement the get_client_data in your subclass.')
+
+    def sentry_error(self, e):
+        raise NotImplementedError('Please implement this method in your subclass.')
+
+    def get_hash_of_bytes(self, item):
+        return mmh3.hash(item)
+
+    def get_hash_of_row(self, row):
+        if isinstance(row, list):
+            items = row
+        elif isinstance(row, Mapping):
+            items = row.items()
+        row_bytes = b','.join([str(v).encode('utf-8') for k, v in items if k not in self.settings.ignore_fields_in_signature_calculation])  # NOQA
+        return self.get_hash_of_bytes(row_bytes)
+
+    def _verify_access_to_s3_bucket(self):
+
+        config = Config(connect_timeout=.5, retries={'max_attempts': 1})
+
+        s3_client = boto3.client('s3', config=config)
+        s3_client.list_objects_v2(Bucket=self.BUCKET_NAME, MaxKeys=1)
+
+    def get_file_from_s3(self, s3key):
+        body = None
+        s3_client = boto3.client('s3')
+        s3fileobj = s3_client.get_object(Bucket=self.BUCKET_NAME, Key=s3key)
+        if 'Body' in s3fileobj:
+            body = s3fileobj['Body'].read()
+            signature = self.get_hash_of_bytes(body)
+            body = body.decode('utf-8')
+        return body, signature
+
+    def report_error_to_slack(self, e, msg='{} The {} failed: {}'):
+        slack_handle_to_ping = f'<{self.settings.slack_handle_to_ping}>' if self.settings.slack_handle_to_ping else''
+        try:
+            msg = msg.format(slack_handle_to_ping, self.JOB_NAME, e)
+        except Exception as e:
+            pass
+        self.slack()
```

Comment:
Possible TypeError. Isn't this call missing the positional argument `text`?

Reply:
@SkamDart Good catch.